pkg/skaffold/build/custom/custom_test.go

```diff
@@ -30,36 +30,61 @@ func TestRetrieveEnv(t *testing.T) {
 	tests := []struct {
 		description   string
 		tag           string
-		pushImages    bool
+		artifact      *latest.CustomArtifact
 		buildContext  string
 		additionalEnv []string
 		environ       []string
 		expected      []string
+		pushImages    bool
+		shouldErr     bool
 	}{
 
 		{
 			description:  "make sure tags are correct",
 			tag:          "gcr.io/image/tag:mytag",
+			artifact:     &latest.CustomArtifact{},
 			environ:      nil,
 			buildContext: "/some/path",
 			expected:     []string{"IMAGE=gcr.io/image/tag:mytag", "IMAGES=gcr.io/image/tag:mytag", "PUSH_IMAGE=false", "BUILD_CONTEXT=/some/path"},
-		}, {
+		},
+		{
 			description:  "make sure environ is correctly applied",
 			tag:          "gcr.io/image/tag:anothertag",
+			artifact:     &latest.CustomArtifact{},
 			environ:      []string{"PATH=/path", "HOME=/root"},
 			buildContext: "/some/path",
-			expected:     []string{"IMAGE=gcr.io/image/tag:anothertag", "IMAGES=gcr.io/image/tag:anothertag", "PUSH_IMAGE=false", "BUILD_CONTEXT=/some/path", "PATH=/path", "HOME=/root"},
-		}, {
+			expected:     []string{"PATH=/path", "HOME=/root", "IMAGE=gcr.io/image/tag:anothertag", "IMAGES=gcr.io/image/tag:anothertag", "PUSH_IMAGE=false", "BUILD_CONTEXT=/some/path"},
+		},
+		{
 			description: "push image is true",
 			tag:         "gcr.io/image/push:tag",
+			artifact:    &latest.CustomArtifact{},
 			pushImages:  true,
 			expected:    []string{"IMAGE=gcr.io/image/push:tag", "IMAGES=gcr.io/image/push:tag", "PUSH_IMAGE=true", "BUILD_CONTEXT="},
-		}, {
+		},
+		{
 			description:   "add additional env",
+			artifact:      &latest.CustomArtifact{},
 			tag:           "gcr.io/image/push:tag",
 			pushImages:    true,
 			additionalEnv: []string{"KUBECONTEXT=mycluster"},
-			expected:      []string{"IMAGE=gcr.io/image/push:tag", "IMAGES=gcr.io/image/push:tag", "PUSH_IMAGE=true", "BUILD_CONTEXT=", "KUBECONTEXT=mycluster"},
+			expected:      []string{"KUBECONTEXT=mycluster", "IMAGE=gcr.io/image/push:tag", "IMAGES=gcr.io/image/push:tag", "PUSH_IMAGE=true", "BUILD_CONTEXT="},
+		},
+		{
+			description: "custom env",
+			artifact: &latest.CustomArtifact{
+				Env: []string{"KEY1=VALUE1"},
+			},
+			tag:      "gcr.io/image/push:tag",
+			expected: []string{"KEY1=VALUE1", "IMAGE=gcr.io/image/push:tag", "IMAGES=gcr.io/image/push:tag", "PUSH_IMAGE=false", "BUILD_CONTEXT="},
+		},
+		{
+			description: "invalid custom env",
+			artifact: &latest.CustomArtifact{
+				Env: []string{"INVALID"},
+			},
+			tag:       "gcr.io/image/push:tag",
+			shouldErr: true,
```

Comment:
I'd like to see a test with overrides, if that's possible.

Reply:
I've re-focused this PR on the buildpacks builder only